### PR TITLE
add getlasterror for windows

### PR DESCRIPTION
**Pull request by anporumb**

### CODE REVIEW

**Comment by jebrando** [comment]
> 	(void)vprintf(format, args);

If you would be so kind since you're in this file change all the tabs to spaces or vise versa.

---
Refers to: src/consolelogger.c:214 in ccab298. [](commit_id = ccab2982f1a6bf74b5b51a276be17dd3638b0748, deletion_comment = False)

File: src/consolelogger.c [hunk 1]
```diff
@@ -7,6 +7,187 @@
...
+        {
+            /*return as is*/
+        }
+    }
+    else
+    {
+        char temp[MESSAGE_BUFFER_SIZE];
+        if (FormatMessage(FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS, NULL, lastError, MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), temp, MESSAGE_BUFFER_SIZE, NULL) == 0)
+        {
+            result = printf_alloc("GetLastError()=0X%08x", lastError);
```
Comment (jebrando) [inline src/consolelogger.c]: >GetLastError() [](start = 35, length = 14)

I really don't like this "GetLastError()=...",  I'd shorten this to be just "0x%x", lastError).  Brevity for the win.

File: src/consolelogger.c [hunk 2]
```diff
@@ -7,6 +7,187 @@
...
+            if ((whereAreThey = strchr(temp, '\r')) != NULL)
+            {
+                *whereAreThey = '\0';
+            }
+            if ((whereAreThey = strchr(temp, '\n')) != NULL)
+            {
+                *whereAreThey = '\0';
+            }
+
+            result = printf_alloc("GetLastError()==0X%08x (%s)", lastError, temp);
```
Comment (jebrando) [inline src/consolelogger.c]: >GetLastError()==0X%08x (%s) [](start = 35, length = 27)

Also here I'd shorten this to be just "0x%x: %s", lastError, temp).

File: src/consolelogger.c [hunk 3]
```diff
@@ -7,6 +7,187 @@
...
+    va_start(args, format);
+
+    switch (log_category)
+    {
+    case AZ_LOG_INFO:
+    {
+        char* printed = vprintf_alloc(format, args);
+        if (printed == NULL)
+        {
+            (void)printf("failure in vprintf_alloc");
```
Comment (jebrando) [inline src/consolelogger.c]: >failure in vprintf_alloc [](start = 26, length = 24)

(void)printf("Info: [FAILED] %s", (options & LOG_LINE) ? "\r\n" : "");

Comment (jebrando) [inline src/consolelogger.c]: Same below so you don't lose the error completely.

---
In reply to: [115879042](https://github.com/Azure/azure-c-shared-utility/pull/76#discussion_r115879042) [](ancestors = 115879042)

File: src/consolelogger.c (hunk 1 above)
Comment (anporumb) [inline src/consolelogger.c]: ok, done

File: src/consolelogger.c (hunk 2 above)
Comment (anporumb) [inline src/consolelogger.c]: this be done too.

File: src/consolelogger.c [hunk 4]
```diff
@@ -7,6 +7,187 @@
...
+                /*return as is*/
+            }
+        }
+    }
+    return result;
+}
+/*this function will use 1x printf (in the happy case) to print an LogError or LogInfo.*/
+/*more than 1x printf / function call can mean intermingled LogErrors in a multithreaded env*/
+/*the function will also attempt to produce some human readable strings for GetLastError*/
+void consolelogger_log(LOG_CATEGORY log_category, const char* file, const char* func, const int line, unsigned int options, const char* format, ...)
```
Comment (anporumb) [inline src/consolelogger.c]: absolutely not necessary.

File: src/consolelogger.c (hunk 3 above)
Comment (anporumb) [inline src/consolelogger.c]: tried.

**Comment by jebrando** [comment]
:shipit:

File: src/consolelogger.c [hunk 5]
```diff
@@ -7,12 +7,194 @@
 #include "azure_c_shared_utility/xlogging.h"
 #include "azure_c_shared_utility/consolelogger.h"
 
+#ifdef _MSC_VER
+#include "windows.h"
```
Comment (jebrando) [inline src/consolelogger.c]: Can we move this to be only around the getlastError() function.  I may want to use the vprintf_alloc function in linux.
